crates/testing/src/byzantine/byzantine_behaviour.rs

```diff
@@ -0,0 +1,314 @@
+            {
+                let mut result = vec![HotShotEvent::DacSend(cert.clone(), sender.clone())];
+                for i in 1..=self.total_views_add_to_cert {
+                    let mut bad_cert = cert.clone();
+                    bad_cert.view_number = cert.view_number + i;
+                    result.push(HotShotEvent::DacSend(bad_cert, sender.clone()));
+                }
+                return result;
+            }
+        }
+        vec![event.clone()]
+    }
+}
+
+/// Information about the hotshot event after parsing
+#[derive(Debug, Clone)]
+pub struct HotShotEventInfo<TYPES: NodeType> {
+    /// Friendly event name
+    pub event_name: String,
+    /// View number
+    pub view_number: u64,
+    /// The cached event
+    pub cached_event: HotShotEvent<TYPES>,
+}
+
+/// View delay configuration
+#[derive(Debug)]
+pub struct ViewDelay<TYPES: NodeType> {
+    /// How many views the node will be delayed
+    pub number_of_views_to_delay: u64,
+    /// A map to received events friendly event name to previous event info
+    pub received_events: HashMap<String, HotShotEventInfo<TYPES>>,
+    /// When leader how many votes were received
+    pub vote_rcv_count: HashMap<String, u64>,
+    /// Number of nodes with stake
+    pub num_nodes_with_stake: u64,
+    /// Specify which view number to stop delaying
+    pub stop_view_delay_at_view_number: u64,
+}
+
+impl<TYPES: NodeType> ViewDelay<TYPES> {
+    fn extract_hotshot_event_info(
+        &self,
+        event: &HotShotEvent<TYPES>,
+    ) -> Option<HotShotEventInfo<TYPES>> {
+        let event_name = &event.to_string();
+        let (hotshot_event_name, view_number) =
+            event_name.split_once('(').unwrap_or((event_name, ""));
+        if view_number.starts_with("view_number=ViewNumber(") {
+            if let Some(start) = view_number.find('(') {
+                if let Some(end) = view_number.find(')') {
+                    let view_number_str = &view_number[start + 1..end];
+
+                    // make sure we can parse the number, if not it will return None
+                    if let Ok(view_num) = view_number_str.parse::<u64>() {
+                        return Some(HotShotEventInfo {
+                            event_name: hotshot_event_name.to_string(),
+                            view_number: view_num,
+                            cached_event: event.clone(),
+                        });
+                    }
+                }
+            }
+        }
+        None
+    }
+
+    fn compare_views(current_view: u64, cached_view: u64, number_of_views_to_delay: u64) -> bool {
+        if current_view <= cached_view {
+            return false;
+        }
+
+        if current_view - cached_view > number_of_views_to_delay {
+            return true;
+        }
+
+        false
+    }
+
+    async fn handle_event(&mut self, event: &HotShotEvent<TYPES>) -> HotShotEvent<TYPES> {
+        if let Some(event_info) = self.extract_hotshot_event_info(event) {
```

Comment:
All this really does is get the view number for the `event` I think it's probably best to just implement a `fn view_number(&self) -> Option<TYPES::Time>` on the `HotshotEvent` enum

I'll looks a lot like `fmt`